File: backpack.pl
#!/usr/bin/perl
use strict;
use warnings FATAL=>'all';

use POSIX qw(strftime);
use Sys::Hostname;


########################################################################
# WHAT IS THAT?
#
# The basic script to synchronize data and make backup!

# VERSION: Sun Jan 10 16:21:32 CET 2016
########################################################################



########################################################################
# Global variables
########################################################################

# BPD dir
my $BPD_DIR = '/home/nobackup/BPD';

# just a shorthand to use only in the completion of @CHOICES
my $REMOTE_DIR = '/mnt/removable/backpack';
# log file
my $LOG_FILE = "$REMOTE_DIR/LOG";

# list of choices available to the user; the format is
# [ { name => NAME_TO_BE_SHOWN,
#     source => SOURCE_DIR,
#     target => TARGET_DIR,
#     local_to_remote => BOOLEAN,
#     bpd => BOOLEAN }, # if this is false or if the key doesn't
#                       # exist, the script asks the user
#   { ... },
#   ...
# ]
my @CHOICES = (

   { name => "bruno",
     source => "/home/bruno",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => 1 },

   { name => "bruno",
     source => "$REMOTE_DIR/bruno",
     target => "/home",
     local_to_remote => '',
     bpd => 1 },


   { name => "storeroom",
     source => "/home/storeroom",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => 1 },

   { name => "storeroom",
     source => "$REMOTE_DIR/storeroom",
     target => "/home",
     local_to_remote => '',
     bpd => 1 },


   { name => "TODO",
     source => "/home/TODO",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => 1 },

   { name => "TODO",
     source => "$REMOTE_DIR/TODO",
     target => "/home",
     local_to_remote => '',
     bpd => 1 },


   { name => "papps",
     source => "/home/papps",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "papps",
     source => "$REMOTE_DIR/papps",
     target => "/home",
     local_to_remote => '',
     bpd => '' },


   { name => "blib",
     source => "/home/blib",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "blib",
     source => "$REMOTE_DIR/blib",
     target => "/home",
     local_to_remote => '',
     bpd => '' },


   { name => "pendingblib",
     source => "/home/pendingblib",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "pendingblib",
     source => "$REMOTE_DIR/pendingblib",
     target => "/home",
     local_to_remote => '',
     bpd => '' },


   { name => "playground",
     source => "/home/playground",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "playground",
     source => "$REMOTE_DIR/playground",
     target => "/home",
     local_to_remote => '',
     bpd => '' },


   { name => "STORAGE: big 01 important",
     source => "/media/bruno/storage/big_01_important",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "STORAGE: big 01 important",
     source => "$REMOTE_DIR/big_01_important",
     target => "/media/bruno/storage",
     local_to_remote => '',
     bpd => '' },

   { name => "STORAGE: big 02 useful",
     source => "/media/bruno/storage/big_02_useful",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "STORAGE: big 02 useful",
     source => "$REMOTE_DIR/big_02_useful",
     target => "/media/bruno/storage",
     local_to_remote => '',
     bpd => '' },

   { name => "STORAGE: big 03 less useful",
     source => "/media/bruno/storage/big_03_less_useful",
     target => $REMOTE_DIR,
     local_to_remote => 1,
     bpd => '' },

   { name => "STORAGE: big 03 less useful",
     source => "$REMOTE_DIR/big_03_less_useful",
     target => "/media/bruno/storage",
     local_to_remote => '',
     bpd => '' },


   { name => "TEST",
     source => "/home/bruno/scripts/newbackpack/testing/test_local",
     target => "/home/bruno/scripts/newbackpack/testing/test_remote",
     local_to_remote => 1,
     bpd => 1 },

   { name => "TEST",
     source => "/home/bruno/scripts/newbackpack/testing/test_remote",
     target => "/home/bruno/scripts/newbackpack/testing/test_local",
     local_to_remote => '' },

);


my $ASCII_ART_TO_LOCAL =<<"END";
                                             ._________________.
  .-----------------.                        | ._____________. |
  |  .-----------.  |                        | |root #       | |
  |[]|           |[]|             |\\         | |             | |
  |  |           |  |             | \\        | |             | |
  |  |           |  |     +-------+  \\       | |             | |
  |  `-----------'  |     |           \\      | !_____________! |
  |     _______ _   |     |           /      !_________________!
  |    |  _    | |  |     +-------+  /          ._[_______]_.
  |    | |_|   | |  |             | /       .___|___________|___
  \\____|_______|_|__|             |/        |::: ____           |
                                            |    ~~~~ [CD-ROM]  |
                                            !___________________!
END

my $ASCII_ART_TO_REMOTE = <<"END";
   ._________________.                        
   | ._____________. |                        .-----------------.
   | |root #       | |                        |  .-----------.  |
   | |             | |              |\\        |[]|           |[]|
   | |             | |              | \\       |  |           |  |
   | |             | |      +-------+  \\      |  |           |  |
   | !_____________! |      |           \\     |  `-----------'  |
   !_________________!      |           /     |     _______ _   |
      ._[_______]_.         +-------+  /      |    |  _    | |  |
  .___|___________|___              | /       |    | |_|   | |  |
  |::: ____           |             |/        \\____|_______|_|__|
  |    ~~~~ [CD-ROM]  |                       
  !___________________!                       
END


########################################################################
# Ask the input data to the user (source and target directories, BPD).
# Return an array: (SOURCE, TARGET, BPD).
#
# Check that the source and target exist.  Ask confirmation.
#
# Return FALSE if the user wants to quit.
########################################################################

sub ask_data_to_user {

   my @choices = @_;

   my $buf;

   # choice of direction

   my $to_remote;
   while (1) {
      print "Choose the direction:\n"
         ."- 0: to remote directory (backup)\n"
         ."- 1: to local directory (restoration)\n"
         ."Choice [0/1/quit]? ";
      $buf = <STDIN>;
      if ($buf =~ m/^\s*+0\s*+$/) {
         print "Choice made: to remote directory.\n";
         $to_remote = 1;
         last;
      } elsif ($buf =~ m/^\s*+1\s*+$/) {
         print "Choice made: to local directory.\n";
         $to_remote = "";
         last;
      } elsif ($buf =~ m/^\s*+quit\s*+$/i) {
         print "Aborted on user request.\n";
         return undef;
      }
      print "Invalid answer.  Try again.\n";
   }

   # choice of source and target dirs

   @choices =
      map { -e $_->{source} ? $_ : () }
      map { (($_->{local_to_remote} and $to_remote)
         or (!$_->{local_to_remote} and !$to_remote)) ? $_ : () } @choices;

   die "$0: *** no source directory available ***\n" unless @choices;

   my $source;
   my $target;
   my $bpd;
   while (1) {
      my $i = 0;
      print "Choose the directory:\n";
      for (@choices) {
         print " - $i: $_->{name}\n";
         $i++;
      }
      print "Choice (one of the leading number, or 'quit')? ";
      $buf = <STDIN>;
      if ($buf =~ m/^\s*+(\d++)\s*+$/ and $1 >= 0 and $1 < scalar @choices) {
         $source = $choices[$1]->{source};
         $target = $choices[$1]->{target};
         if (exists $choices[$1]->{bpd}) {
            $bpd = $choices[$1]->{bpd};
         } else {
            $bpd = '';
         }
         last;
      } elsif ($buf =~ m/^\s*+quit\s*+$/i) {
         print "Aborted on user request.\n";
         return undef;
      }
      print "Invalid answer.  Try again.\n";
   }

   die "$0: *** source directory '$source' doesn't exist ***\n"
      unless -d-r-w-x $source;

   die "$0: *** target directory '$target' doesn't exist ***\n"
      unless -d-r-w-x $target;

   # ask confirmation

   while (1) {
      my $color = $to_remote ? '32' : '31';
      print "Choice made:\n"
         ."\033[$color;1mSource:\n$source\n"
         .($to_remote ? $ASCII_ART_TO_REMOTE : $ASCII_ART_TO_LOCAL)
         .sprintf("% 65s\n", "Target:")
         .sprintf("% 65s\033[0m\n", $target);
      if ($to_remote) {
         print "Is this correct [y/n]? ";
      } else {
         print "Type 'erase local data' to continue (or 'quit'): ";
      }
      $buf = <STDIN>;
      if ($to_remote and $buf =~ m/^\s*+y\s*+$/) {
         last;
      } elsif ($to_remote and $buf =~ m/^\s*+n\s*+$/i) {
         print "Aborted on user request.\n";
         return undef;
      } elsif (!$to_remote and $buf =~ m/^\s*+quit\s*+$/i) {
         print "Aborted on user request.\n";
         return undef;
      } elsif (!$to_remote and $buf =~ m/^\s*+erase\s++local\s++data\s*+$/i) {
         last;
      }
      print "Invalid answer.  Try again.\n";
   }

   # ask for BPD

   unless ($bpd) {
      while (1) {
         print "Do you want to use BPD [y/n/quit]? ";
         $buf = <STDIN>;
         if ($buf =~ m/^\s*+y\s*+$/) {
            print "Choice made: use BPD.\n";
            $bpd = 1;
            last;
         } elsif ($buf =~ m/^\s*+n\s*+$/) {
            print "Choice made: no BPD.\n";
            $bpd = '';
            last;
         } elsif ($buf =~ m/^\s*+quit\s*+$/i) {
            print "Aborted on user request.\n";
            return undef;
         }
         print "Invalid answer.  Try again.\n";
      } # while
   } # unless

   return ($source, $target, $bpd);

}


########################################################################
# Build the rsync options.  Return an array with all the options, to
# be used with the system() command.
########################################################################

sub build_rsync_options {

   my $source = shift;
   my $target = shift;
   my $bpd = shift;
   my $timestamp = shift;

   my @rsync_options = (
      "--modify-window", "1", # loose comparision of timestamp
      "--recursive", # descend into directories
      "--times", # update timestamps
      "--perms", # update permissions
      "--owner", # update owner (only if rsync is run as superuser)
      "--group", # update group (idem)
      "--links", # copy symlinks as symlinks
      "--devices", # recreate character and block device files (only if run as superuser)
      "--specials", # recreate special files (socket, fifos (only if run as superuser))
      "--delete", # delete extraneous files from dest dirs (otherwise, these are kept)
      "--human-readable", # output numbers in a human-readable format
      "-v", "-v", # verbose, and more verbose
      "--progress", # show progress during transfer
      "--log-file", "backup.log", # save a log file
   );

   if ($bpd) {
      die "$0: *** BPD directory '$BPD_DIR' doesn't exist ***\n"
         unless -d-r-w-x $BPD_DIR;
      my $bpd_dir = sprintf('/home/nobackup/BPD/%s/', $timestamp);
      die "$0: *** BPD directory '$bpd_dir' already exists ***\n"
         if -e $bpd_dir;
      push @rsync_options, '--backup', '--backup-dir', $bpd_dir;
   }

   # dirname = dir name without the parent dirs
   (my $dirname = $source) =~ s{^([^/]*+/)*+}{};
   push @rsync_options, '--exclude', "/$dirname/.*";

   push @rsync_options, $source, $target;

   return @rsync_options;

}


########################################################################
# print log
########################################################################

sub write_log {

   my $file = shift;
   my $timestamp = shift; # if undef, just write 'ok'
   my $source = shift;
   my $target = shift;
   my $msg = shift;
   my @options = @_;

   open my $fh, ">>", $file or die "$0: *** can't open $file ***\n";

   if ($timestamp) {
      #print "time: $timestamp\n";
      #print "msg: $msg\n";
      #print "src: $source\n";
      #print "targ: $target\n";
      #print "opt: @options\n";
      my $host = hostname();
      print $fh "$timestamp ($host, $msg): $source --> $target (@options)\n";
   } else {
      print $fh "... ok\n";
   }

   close $fh or die "$0: *** can't close $file ***\n";

}



########################################################################
# main()
########################################################################

sub main {

   # only root can do that!
   my @choices;
   if ($ENV{LOGNAME} eq 'root') {
      @choices = map { $_->{name} !~ m/^TEST\b/ ? $_ : () } @CHOICES;
   } else {
      @choices = map { $_->{name} =~ m/^TEST\b/ ? $_ : () } @CHOICES;
   }

   # get the data
   my ($source, $target, $bpd) = ask_data_to_user(@choices);
   return unless $source and $target;

   my $timestamp = strftime("%Y%m%d_%H%M%S", gmtime);

   # build the command
   my @rsync_options = build_rsync_options($source, $target, $bpd, $timestamp);
   return unless @rsync_options;

   # print summary

   while (1) {
      print '-'x80, "\n";
      print 
         "Summary (check SOURCE, TARGET and BPD):\n"
         ."   \033[1;37mSource:\033[0m  $source\n"
         ."   \033[1;37mTarget:\033[0m  $target\n"
         ."   \033[1;37mUse BPD:\033[0m ".($bpd ? 'yes' : 'no')."\n"
         ."   \033[1;37mCommand:\033[0m rsync @rsync_options\n"
         ."   \033[1;37mNote:\033[0m No $source/.* files or dirs will be saved!\n";
      print '-'x80, "\n";
      print "Do you want to continue (last chance to stop) [y/n]? ";
      my $buf = <STDIN>;
      if ($buf =~ m/^\s*+y\s*+$/) {
         print "Ready to go.\n";
         last;
      } elsif ($buf =~ m/^\s*+n\s*+$/) {
         print "Aborted on user request.\n";
         return;
      }
      print "Invalid answer.  Try again.\n";
   }

   print "Enter some message for the log (home, library...): ";
   my $msg = <STDIN>;
   chomp $msg;
   write_log($LOG_FILE, $timestamp, $source, $target, $msg, @rsync_options);

   # run the command
   print "Running the command...\n";
   system('rsync', @rsync_options)
      and die "$0: *** error when running rsync ***\n$!";

   write_log($LOG_FILE);

}

main();
print "done!\n";

File: backpack.py
import sys, os, time, socket, subprocess, collections, parse_rsync_log

## parameters ##########################################################

Data = collections.namedtuple("Data", "local_path force_bpd")

TEST_MODE = False

if TEST_MODE:
    REMOTE_DIR = '/tmp/backpack/remote'
    BPD_DIR = '/tmp/backpack/BPD';
    DATA = [
        Data('/tmp/backpack/local/somedir', True),
        Data('/tmp/backpack/local/someotherdir', False)
    ]

else:
    REMOTE_DIR = '/mnt/removable/backpack'
    BPD_DIR = '/home/nobackup/BPD';
    DATA = [
        Data('/home/bruno', True),
        Data('/home/storeroom', True),
        Data('/home/blib', True),
        Data('/home/pendingblib', True),
        Data('/home/papps', False),
        Data('/home/playground', False),
        Data('/home/TODO', False),
        Data('/media/bruno/storage/big_01_important', False),
        Data('/media/bruno/storage/big_02_useful', False),
        Data('/media/bruno/storage/big_03_less_useful', False),
    ]


########################################################################
#                  NO MODIFICATION REQUIRED BELOW THIS LINE
########################################################################

## constants ###########################################################

TO_REMOTE = 1
TO_LOCAL = 2

ASCII_TO_LOCAL = r"""
                                             ._________________.
  .-----------------.                        | ._____________. |
  |  .-----------.  |                        | |root #       | |
  |[]|           |[]|             |\         | |             | |
  |  |           |  |             | \        | |             | |
  |  |           |  |     +-------+  \       | |             | |
  |  `-----------'  |     |           \      | !_____________! |
  |     _______ _   |     |           /      !_________________!
  |    |  _    | |  |     +-------+  /          ._[_______]_.
  |    | |_|   | |  |             | /       .___|___________|___
  \____|_______|_|__|             |/        |::: ____           |
                                            |    ~~~~ [CD-ROM]  |
                                            !___________________!
"""

ASCII_TO_REMOTE = r"""
   ._________________.                        
   | ._____________. |                        .-----------------.
   | |root #       | |                        |  .-----------.  |
   | |             | |              |\        |[]|           |[]|
   | |             | |              | \       |  |           |  |
   | |             | |      +-------+  \      |  |           |  |
   | !_____________! |      |           \     |  `-----------'  |
   !_________________!      |           /     |     _______ _   |
      ._[_______]_.         +-------+  /      |    |  _    | |  |
  .___|___________|___              | /       |    | |_|   | |  |
  |::: ____           |             |/        \____|_______|_|__|
  |    ~~~~ [CD-ROM]  |                       
  !___________________!                       
"""

## common functions ####################################################

def print_in_red(msg):
    print("\033[31;1m%s\033[0m" % msg)

def print_in_yellow(msg):
    print("\033[33;1m%s\033[0m" % msg)

def ask_for_direction():
    while True:
        buf = input("Choose direction [l = to local, r = to remote, quit]: ")
        buf = buf.strip().lower()
        if buf == 'l':
            while True:
                print("\033[1;31m"+ASCII_TO_LOCAL+"\033[0m")
                buf = input("Type `erase local data' to confirm (or quit): ")
                buf = buf.strip().lower()
                if buf == 'erase local data':
                    return TO_LOCAL
                elif buf == 'quit':
                    raise QuitException
                print("Invalid answer.  Try again.")
        elif buf == 'r':
            while True:
                print("\033[1;32m"+ASCII_TO_REMOTE+"\033[0m")
                buf = input("Is this correct [y/n]? ")
                buf = buf.strip().lower()
                if buf == 'y':
                    return TO_REMOTE
                elif buf == 'quit' or buf == 'n':
                    raise QuitException
                print("Invalid answer.  Try again.")
        elif buf == 'quit':
            raise QuitException
        print("Invalid answer.  Try again.")

def get_bpd_dir(bpd_base_dir, timestamp):
    if not os.path.exists(bpd_base_dir):
        raise FileNotFoundError("%s not found" % bpd_base_dir)
    if not os.path.isdir(bpd_base_dir):
        raise FileNotFoundError("%s not a directory" % bpd_base_dir)
    path = os.path.join(bpd_base_dir, timestamp)
    if os.path.exists(path):
        raise FileExistsError("%s exists" % path)
    return path

def confirm(msg, default=None):
    while True:
        buf = input(msg).strip()
        if buf == 'y':
            return True;
        elif buf == 'n':
            return False
        elif buf == '' and default != None:
            return default
        else:
            print("Invalid answer.  Please try again.")

def ask_location():
    while True:
        buf = input("Enter your location: ")
        buf = buf.strip()
        if buf:
            return buf
        print("Invalid answer.  Try again.")

## exceptions ##########################################################

class QuitException(Exception):
    pass

class FileExistsError(Exception):
    def __init__(self, msg):
        super().__init__(msg)

########################################################################

class SynchroItem():

    def __init__(self, name, local_dir, remote_dir, direction, force_bpd,
            log_file, bpd_dir):
        assert direction in (TO_LOCAL, TO_REMOTE)
        self.source_dir = os.path.join(local_dir if direction == TO_REMOTE
            else remote_dir, name)
        self.dest_dir = local_dir if direction == TO_LOCAL else remote_dir
        self.direction = direction
        self.remote_dir = remote_dir
        self.force_bpd = force_bpd
        self.log_file = log_file
        self.bpd_dir = bpd_dir
        self._use_bpd = False
        self._ok_to_synchronize = False
        self._parse_rsync_log = False
        self._ask()
        self._rsync_launcher = None
        if self.ok_to_synchronize:
            self._rsync_launcher = RsyncLauncher(source_dir=self.source_dir,
                dest_dir=self.dest_dir,
                log_file=self.log_file,
                bpd_dir=self.bpd_dir if self._use_bpd else None,
                parse_log=self._parse_rsync_log,
                remote_dir=self.remote_dir,
                direction=self.direction)

    @property
    def ok_to_synchronize(self):
        return self._ok_to_synchronize

    @property
    def use_bpd(self):
        return self._use_bpd

    @property
    def parse_rsync_log(self):
        return self._parse_rsync_log

    @property
    def rsync_launcher(self):
        return self._rsync_launcher

    def __str__(self):
        return "%s -> %s" % (self.source_dir, self.dest_dir)

    def _ask(self):
        # warning
        if not os.path.exists(self.source_dir):
            print_in_red("The source dir `%s' " "doesn't exist. Skipping."
                % self.source_dir)
            input("Press Enter to continue...")
            return
        target_dir = os.path.join(self.dest_dir,
            os.path.basename(self.source_dir))
        if not os.path.exists(target_dir):
            print_in_red("Warning: The target directory `%s' doesn't exist."
                % target_dir)
        # ask
        if not confirm("Synchronize `\033[1m%s\033[0m' [Y/n]? " %
                self.source_dir, True):
            print("Skipping `%s'." % self.source_dir)
            return
        self._ok_to_synchronize = True
        # use bpd?
        if self.force_bpd:
            self._use_bpd = True
        elif confirm("Use BPD [y/N]? ", False):
            self._use_bpd = True
        else:
            self._use_bpd = False
        # create md5?
        if confirm("Create md5 sums from log file [Y/n]?", True):
            self._parse_rsync_log = True

    def write_log(self, path, location, success=False):
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime())
        if success:
            state = "success"
        else:
            state = "not yet executed"
        open(path, 'a').write("%s @ %s: %s -> %s (%s)\n" % \
            (timestamp, location, self.source_dir, self.dest_dir, state))

########################################################################

class RsyncLauncher():

    default_args = (
        "rsync",
        # "--modify-window", "1", # loose comparison of timestamp
        "--recursive", # descend into directories
        "--times", # update timestamps
        "--perms", # update permissions
        "--owner", # update owner (only if rsync is run as superuser)
        "--group", # update group (idem)
        "--links", # copy symlinks as symlinks
        "--devices", # recreate character and block device files (only if run
                     # as superuser)
        "--specials", # recreate special files (socket, fifos (only if run as
                      #superuser))
        "--delete", # delete extraneous files from dest dirs (otherwise, these
                    # are kept)
        "--human-readable", # output numbers in a human-readable format
        #"-v", # verbose
        #"-v", # and more verbose
        "--progress", # show progress during transfer
    )

    def __init__(self, source_dir, dest_dir, log_file, bpd_dir=None,
            parse_log=False, remote_dir=None, direction=0):
        self.args = list(RsyncLauncher.default_args)
        self.source_dir = source_dir
        self.dest_dir = dest_dir
        self.log_file = log_file
        self.bpd_dir = bpd_dir
        self.parse_log = parse_log
        self.remote_dir=remote_dir
        self.md5_file_written_to = False
        self.direction = direction
        if self.parse_log:
            assert self.remote_dir
            assert direction in (TO_LOCAL, TO_REMOTE)
        # more arguments for rsync
        self.args.extend(('--exclude', '/%s/.*' %
            os.path.basename(self.source_dir)))
        assert self.log_file
        self.args.extend(("--log-file", self.log_file))
        if self.bpd_dir:
            self.args.extend(('--backup', '--backup-dir', self.bpd_dir))
        self.args.append(self.source_dir)
        self.args.append(self.dest_dir)

    def run(self):
        subprocess.run(self.args, check=True)
        if self.parse_log:
            if self.direction == TO_REMOTE:
                log_dir = os.path.join(self.remote_dir, 'logs')
            else:
                log_dir = 'logs'
            if not os.path.exists(log_dir):
                os.mkdir(log_dir)
            elif not os.path.isdir(log_dir):
                raise RuntimeError("`%s' is not a directory" % log_dir)
            output_file = os.path.join(log_dir, os.path.basename(self.log_file))
            output_file += ".md5"
            ans = parse_rsync_log.run(log_file=self.log_file,
                source_dir=os.path.dirname(self.source_dir),
                dest_dir=self.dest_dir,
                output_file=output_file)
            if ans:
                self.md5_file_written_to = output_file

    def __str__(self):
        return ' '.join(self.args)


########################################################################

if __name__ == '__main__':

    if not (TEST_MODE or os.getenv('LOGNAME', '') == 'root'):
        raise RuntimeError("you're not testing and you're not root!")
    if not os.path.exists(REMOTE_DIR):
        raise FileNotFoundError(REMOTE_DIR)

    # variables
    try:
        direction = ask_for_direction()
    except QuitException:
        print("Aborted on user request.")
        sys.exit(0)
    timestamp = time.strftime("%Y%m%d_%H%M%S")
    bpd_dir = get_bpd_dir(BPD_DIR, timestamp)
    rsync_log_dir = 'logs'
    if not os.path.exists(rsync_log_dir):
        os.mkdir(rsync_log_dir)
    elif not os.path.isdir(rsync_log_dir):
        raise RuntimeError("`%s' is not a directory" % rsync_log_dir)
    rsync_log_file = os.path.join(rsync_log_dir, "backup_%s_%%s.log" %
        timestamp)
    backpack_log_file = os.path.join(REMOTE_DIR, 'LOG')
    hostname = socket.gethostname()
    if not hostname:
        raise RuntimeError("no hostname found")
    location = ask_location()
    location = "%s (%s)" % (hostname, location)

    items = []
    log_count = 0
    for data in DATA:
        items.append(SynchroItem(
            name=os.path.basename(data.local_path),
            local_dir=os.path.dirname(data.local_path),
            remote_dir=REMOTE_DIR,
            direction=direction,
            force_bpd=data.force_bpd,
            log_file=rsync_log_file % chr(log_count+97),
            bpd_dir=bpd_dir))
        log_count += 1

    count = 0
    for item in items:
        if item.ok_to_synchronize:
            count += 1

    # confirmation
    if not count:
        print("Nothing to synchronize!")
        sys.exit(0)
    print('='*72)
    print("I will run the following commands:")
    for item in items:
        if item.ok_to_synchronize:
            print("\033[1m%s:\033[0m" % item.source_dir)
            print(str(item.rsync_launcher))
    print("Summary:")
    bpd_yes = "\033[1;32mBPD\033[0m"
    bpd_no = "\033[1;31mBPD\033[0m"
    md5_yes = "\033[1;32mmd5\033[0m"
    md5_no = "\033[1;31mmd5\033[0m"
    for item in items:
        if item.ok_to_synchronize:
            print("- \033[1m%s -> %s \033[0m (%s, %s)" % (
                item.source_dir,
                item.dest_dir,
                bpd_yes if item.use_bpd else bpd_no,
                md5_yes if item.parse_rsync_log else md5_no))
        else:
            print("- \033[1;31m%s\033[0m (not synchronized)" % (item.source_dir))
    while True:
        buf = input("Are you ok (type `ok' or `q[uit]')? ")
        buf = buf.lower().strip()
        if buf == "ok":
            break
        elif buf in ("q", "quit"):
            print("You're not ok. Nothing done!")
            sys.exit(0)
        print("I don't understand. Try again.")

    for item in items:
        if item.ok_to_synchronize:
            print("\033[1m" + "="*72 + "\033[0m")
            item.write_log(backpack_log_file, location, False)
            print("Running: " + str(item.rsync_launcher))
            item.rsync_launcher.run() 
            item.write_log(backpack_log_file, location, True)

    print("List of md5 files:")
    for item in items:
        if item.ok_to_synchronize and item.rsync_launcher.md5_file_written_to:
            print("- %s" % item.rsync_launcher.md5_file_written_to)

    print('done!')

